test/unittest/tests/test_futex.cpp

```diff
@@ -0,0 +1,295 @@
+     * out then calling wake so treat as an error.
+     */
+    TimeoutWithUnexpecedValue,
+
+    /*
+     * This could because this value was never present in the wakeup value list, or more waiters
+     * than expected were awoken with the same value.
+     */
+    WakeWithUnexpectedValue,
+
+    /* An unexpected non-spurious wake up is present. This is almost certainly a bug in the test. */
+    UnexpectWakeNonSpuriousWakeup,
+
+    /* Unexpected timeouts are a different type of soft failure to spurious wakeups. */
+    UnexpectedTimeouts,
+};
+
+bool
+OutcomeNotFailure(Outcome result)
+{
+    return result == Outcome::Expected || result == Outcome::Acceptable;
+}
+
+ostream &
+operator<<(ostream &out, const Outcome &result)
+{
+#ifdef R_
+#error "R_ already defined"
+#endif
+
+#define R_(r)      \
+    case r:        \
+        out << #r; \
+        break
+
+    switch (result) {
+        R_(Expected);
+        R_(Acceptable);
+        R_(Error);
+        R_(RetZeroWithErrno);
+        R_(TimeoutWithUnexpecedValue);
+        R_(WakeWithUnexpectedValue);
+        R_(UnexpectWakeNonSpuriousWakeup);
+        R_(UnexpectedTimeouts);
+    }
+
+#undef R_
+    return out;
+}
+/*
+ * check_outcomes --
+ *     Check completed waiters to see if they meet expectations.
+ *
+ * There must be a SEPARATE wake up value for each and every waiter that is expected to be awoken.
+ */
+Outcome
+check_outcomes(
+  const vector<Waiter> &waiters, const unsigned int timeouts, const vector<futex_word> &wake_vals)
+{
+    list<futex_word> wake_values(wake_vals.begin(), wake_vals.end());
+    unsigned int timedout = 0;
+
+    /*
+     * Loop over the waiters and inspect their state.
+     *
+     * Note: spurious wakeups (no error and expected == wake up value) are detected after the loop
+     * exits so there are is no test in the loop.
+     */
+    for (auto &&w : waiters) {
+        if (w.ret == -1) {
+            if (w.eno != ETIMEDOUT)
+                return Outcome::Error;
+            else if (w.val_on_wake != w.expected)
+                return Outcome::TimeoutWithUnexpecedValue;
+            timedout++;
+        } else if (w.eno != 0)
+            return Outcome::RetZeroWithErrno;
+        else if (wake_values.empty() && w.val_on_wake != w.expected)
+            return Outcome::UnexpectWakeNonSpuriousWakeup;
+        else if (w.val_on_wake != w.expected) {
+            /* This is a non-spurious wake up. */
+            auto match = find(wake_values.begin(), wake_values.end(), w.val_on_wake);
+            if (match == wake_values.end())
+                return Outcome::WakeWithUnexpectedValue;
+            wake_values.erase(match);
+        }
+    }
+
+    if (timedout > timeouts) {
+        return UnexpectedTimeouts;
+    }
+
+    if (wake_values.empty() && timedout == timeouts)
+        return Outcome::Expected;
+
+    /*
+     * Spurious wake ups detected: while this is not an error, is not a consider successful.
```

Comment:
Nitpick: `is not a consider successful` wording is awkward

Reply:
This is still an issue after the latest commit.

Reply:
Actually would you mind proof-reading the comments in this file generally? I noticed some more grammar issues this time around.